### PR TITLE
Fix #1515: signal error if output parameter doesn't match

### LINKED ISSUE #1515: Handling of output parameter for worklet process()?
Let's say you have a very simple `AudioWorklet` having a single input and a single output with no `AudioParams`.  Let's say make a mistake and have

```
process(inputs, outputs, params) {
  let input = inputs[0];
  let output = outputs[0];
  output.fill(input[0][0]);
  return true;
}
```
This accidentally modifies `outputs[0]` to be an array of floats.  But it's supposed to be a `sequence<Float32Array>`.

How is the worklet supposed to handle this?  I'm kind of thinking this should cause an error to be thrown.  I guess that also implies that if `output` isn't exactly `sequence<sequence<Float32Array>>` an error is thrown.  Similarly, if the length of the `Float32Array` isn't 128, it's an error.

(In case it matters, I actually did this by accident.)

### PR DESCRIPTION
**Pull request by rtoy**
If the output parameter on return from process() doesn't match the
structure on calling process(), throw an error.

***
<a href="https://pr-preview.s3.amazonaws.com/rtoy/web-audio-api/pull/1518.html" title="Last updated on Apr 13, 2018, 6:59 PM GMT (79a6706)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WebAudio/web-audio-api/1518/cf8b377...rtoy:79a6706.html" title="Last updated on Apr 13, 2018, 6:59 PM GMT (79a6706)">Diff</a>

### CODE REVIEW

**Comment by rtoy** [comment]
Can't merge this until #1517 lands (missing links), but the text should e correct.

File: index.bs [hunk 1]
```diff
@@ -9815,6 +9815,24 @@ The method is invoked with the following arguments:
...
+
+		1. Check <var>outputs_out</var>:
+			1. <var>outputs_out</var> must be a sequence whose length is the same as the length of <var>outputs_in</var>.
+			1. <var>outputs_out[n]</var> must be a sequence whose length is the same the length of <var>outputs_in[n]</var>.
+			1. <var>outputs_out[n][m]</var> must be a {{Float32Array}} whose length is the same the length of <var>outputs_in[n][m]</var>.
+		1. If all checks above pass, exit this algorithm.
+		1. Otherwise some check failed and perform the following steps:
+			1. Deactivate the processor, resulting in silent output from the processor for all subsequent renderings.
+			1. Throw an exception
+			1. Fire the {{AudioWorkletNode}}'s {{onprocessorerror}} event.
```
Comment (padenot) [inline index.bs]: This is best worded as "queue a task on some event loop to fire the `"processorerror"` event".

File: index.bs [hunk 2]
```diff
@@ -9815,6 +9815,24 @@ The method is invoked with the following arguments:
...
+		<code>process()</code>.
+
+		1. Check <var>outputs_out</var>:
+			1. <var>outputs_out</var> must be a sequence whose length is the same as the length of <var>outputs_in</var>.
+			1. <var>outputs_out[n]</var> must be a sequence whose length is the same the length of <var>outputs_in[n]</var>.
+			1. <var>outputs_out[n][m]</var> must be a {{Float32Array}} whose length is the same the length of <var>outputs_in[n][m]</var>.
+		1. If all checks above pass, exit this algorithm.
+		1. Otherwise some check failed and perform the following steps:
+			1. Deactivate the processor, resulting in silent output from the processor for all subsequent renderings.
+			1. Throw an exception
```
Comment (padenot) [inline index.bs]: Which one?

File: index.bs [hunk 3]
```diff
@@ -9815,6 +9815,24 @@ The method is invoked with the following arguments:
...
+		<code>outputs</code> parameter on return from
+		<code>process()</code>.
+
+		1. Check <var>outputs_out</var>:
+			1. <var>outputs_out</var> must be a sequence whose length is the same as the length of <var>outputs_in</var>.
+			1. <var>outputs_out[n]</var> must be a sequence whose length is the same the length of <var>outputs_in[n]</var>.
+			1. <var>outputs_out[n][m]</var> must be a {{Float32Array}} whose length is the same the length of <var>outputs_in[n][m]</var>.
+		1. If all checks above pass, exit this algorithm.
+		1. Otherwise some check failed and perform the following steps:
+			1. Deactivate the processor, resulting in silent output from the processor for all subsequent renderings.
```
Comment (padenot) [inline index.bs]: Can we have this concept defined somewhere? It's going to be useful in other cases for sure. I'm thinking of an exception thrown from the worklet, or an UA might wish to deactivate a processor if it's using too much CPU, or the like.

File: index.bs [hunk 4]
```diff
@@ -9815,6 +9815,24 @@ The method is invoked with the following arguments:
 	<a>computedNumberOfChannels</a> only when the node has single
 	output.
 
+	<div algorithm="check-process-outputs">
+		Let <var>outputs_in</var> be the <code>outputs</code> passed
+		to <code>process()</code>.  let <var>outputs_out</var> be the
```
Comment (padenot) [inline index.bs]: s/let/Let/

File: index.bs [hunk 5]
```diff
@@ -9815,6 +9815,24 @@ The method is invoked with the following arguments:
...
+		<code>outputs</code> parameter on return from
+		<code>process()</code>.
+
+		1. Check <var>outputs_out</var>:
```
Comment (padenot) [inline index.bs]: I find this a bit weird. Can you lay out the steps normally with `if` and `and` statements?

File: index.bs (hunk 2 above)
Comment (rtoy) [reply index.bs]: F2F resolutio: InvalidModificationError

File: index.bs (hunk 4 above)
Comment (rtoy) [reply index.bs]: Done

File: index.bs (hunk 5 above)
Comment (rtoy) [reply index.bs]: Done, I think.  Is this better?

File: index.bs (hunk 3 above)
Comment (rtoy) [reply index.bs]: Done.

File: index.bs (hunk 2 above)
Comment (rtoy) [reply index.bs]: Done

File: index.bs (hunk 1 above)
Comment (rtoy) [reply index.bs]: Done.  Is this good enough for now?

File: index.bs [hunk 6]
```diff
@@ -9885,6 +9885,41 @@ The method is invoked with the following arguments:
...
+			1. and <var>outputs_out</var>\([n]\) is a
+				sequence whose length is equal to the
+				length of <var>outputs_in</var>\([n]\)
+				for each valid \(n\).
+			1. and <var>outputs_out</var>[n][m] is a
+				{{Float32Array}} whose length is equal
+				to the length of
+				<var>outputs_in</var>\([n][m]\) for
+				each valid \(m\).
+		1. Then exit this algorithm.
```
Comment (hoch) [inline index.bs]: Can this item be rolled up to l.9894?

File: index.bs [hunk 7]
```diff
@@ -9885,6 +9885,41 @@ The method is invoked with the following arguments:
...
+		1. Otherwise some check failed and perform the following steps:
```
Comment (hoch) [inline index.bs]: I think we can remove `some check failed and`.

**Comment by rtoy** [comment]
Oh, sorry. I think this is on hold because #1515 has some [additional comments](https://github.com/WebAudio/web-audio-api/issues/1515#issuecomment-384078998) that we might be able to use `Object.freeze`. If so, this PR would be obsolete.

**Comment by rtoy** [comment]
Closing this.  We can do this again when #1515 is clarified.

**Pull request closed without merging**